Overall code refactoring (ES Modules) + Update gulp and other dev dependencies

Raruto opened the pull request:
WIP, integrates and superseed: https://github.com/g3w-suite/g3w-client/pull/34

- overall replace function prototypes with native javascript classes
- overall replace commonjs `require()` with native `import()` modules
- refactor `gulpfile.js` and update to support gulp v4
- refactor `config.template.js` (rename some variables, add comments and use fewer nested structures)
- remove thousands of unused dev dependencies
- remove `preinstall` hook (no more `npm-force-resolutions`)
- rename some npm scripts to use a more common nomeclature (see: `npm run dev` and `npm run build`)
- remove gulp task `g3w-admin:client_only`
- rename some gulp tasks to use shorter names
- remove useless `path.join` within gulpfile.js (nowdays filesystem scripts still works without this hack on Windows OS)
- upgrade `babel` to version 7
- add `esmify` package to enable support of ES6 modules within browserify
- remove `.babelrc` and `.browserslistrc` files (related config are now listed within the package.json)
- rename index.html templates (see: `index.dev.html` and `index.prod.html`)
- split i18n config into multiple language files
- install and configure `eslint` 8
- apply airbnb style rules (see: [eslint-config-airbnb-base](https://www.npmjs.com/package/eslint-config-airbnb-base))
- avoid calling the `super ()` constructor when not strictly necessary within javascript classes
- avoid importing local files by relatively traversing project directories (eg. `import foo from '../../../foo'`)
- remove useful `boostrap()` function closure within main.js
- remove useful `test` property within api.js
- remove useful `G3W_VERSION` and `production` variable within applicationservice.js
- updata update `version` to 4.0.0. within package.json
- create `gui/component` folder and move related gui components inside that
- replace `_.forEach` with `Object.entries().forEach`
- remove `src/g3w-ol/src` folder (replaced with `src/g3w-ol`)

---

**To be performed (not yet completed)**

- make use of vue single component (consolidate `*.js` and `*.html` files into a single `.vue` file)

Raruto committed:
update gulp and other dev dependencies

- refactor gulpfile.js
- refactor config.template.js
- remove unusued dev dependencies

volterra79 committed:
first refactoring using class instead prototype and use es6 module instead commonjs

Raruto committed:
Merge branch 'gulp-4' into migration_v3_v4

Raruto committed:
Update package-lock.json

volterra79 committed:
aligned engines in package.json from dev branch

volterra79 committed:
aligned i18n  with last dev changes

volterra79 committed:
fixed issue on refactoring prototype methods to class methods

Raruto committed:
Merge branch 'migration_v3_v4' of https://github.com/g3w-suite/g3w-client into migration_v3_v4

Raruto committed:
replace "function() =" within classes

Raruto committed:
Babel 7 test

- remove .babelrc
- remove .browserlisrc

probably not working due to: https://github.com/vuejs/vueify/issues/237

Raruto committed:
remove require() statements within `/src` files

Raruto committed:
try again with babel 6

Raruto committed:
include "node_modules" folder in babelify() transform

Raruto committed:
fix some relative imports

Raruto committed:
rename "clean" scripts

Raruto committed:
globally replace "){" with ") {"

Raruto committed:
fix missing "admin" script within package.json

Raruto committed:
vueify + esmify + babel 7

Raruto committed:
remove global option

Raruto committed:
remove vue3 "ref"

volterra79 committed:
Fixed issue on refactoring prototype to class

volterra79 committed:
Fixed issue on geolocation, meauser control (length, area) and aligned with dev print changes

volterra79 committed:
refactored nominatim control.

Raruto commented on gulpfile.js:
```diff
@@ -160,8 +160,8 @@ gulp.task('browserify', function(done) {
-      // sourceMaps: true, 
-      // ignore: [/\/node_modules\/(?!ol\/)/],
+      // sourceMaps: true,
+      ignore: [/\/node_modules\//],
```
@volterra79  is there a particular reason why you did this?

Raruto commented on gulpfile.js:
```diff
@@ -170,7 +170,6 @@ gulp.task('browserify', function(done) {
   const bundle = function() {
     return bundler.bundle()
       .on('error', function(err) {
-        console.log(err);
```
@volterra79 can i safely restore that line?

Raruto replied:
@volterra79 can I mark it as resolved?

volterra79 replied:
yes

Raruto commented on gulpfile.js:
```diff
@@ -315,7 +314,7 @@ gulp.task('html:compiletemplate', function() {
 
 gulp.task('browser-sync', function() {
-  const port = g3w.port ?? 3000;
+  const port = g3w.localServerPort ?? 3000;
```
@volterra79  If you think this might give some problems with the current `config.js` (having to develop between the v4 and v3.x branches in the future), I can add some checks on the old name of theese "breaking" properties at the beginning of the file `gulpfile.js`. In this way (with the appropriate comments) the intent should be clearer.

Raruto replied:
https://github.com/g3w-suite/g3w-client/pull/35/commits/9f702d65178b80081fe15accb5c6317a7d11ad91 should take care of this

Raruto replied:
@volterra79 can I mark it as resolved?

volterra79 replied:
yes

Raruto commented on gulpfile.js:
```diff
@@ -327,8 +326,8 @@ gulp.task('browser-sync', function() {
           if (req.url.indexOf('plugin.js') > -1) rootUrl = req.url;
           else rootUrl = req.url.split('?')[0];
-          for (let i in g3w.proxy.routes) {
-            if (rootUrl.indexOf(g3w.proxy.routes[i]) > -1) {
```
@volterra79 the same goes here, I think `proxy.routes` name is more appropriate than a generic `proxy.urls`

Raruto replied:
https://github.com/g3w-suite/g3w-client/pull/35/commits/9f702d65178b80081fe15accb5c6317a7d11ad91 should take care of this

Raruto replied:
@volterra79 can I mark it as resolved?

volterra79 replied:
Yes

volterra79 committed:
alligned printconfig.js with pull request #38

volterra79 committed:
fixed BingLayer refactoring bug

volterra79 committed:
added romanian traslation to print section

Raruto committed:
simplify gulpfile.js

- remove `build_all` conditional
- add some backcomp to previous config.template.js (v3)
- remove useless `path.join` (nowdays filesystem scripts still works without this hack on Windows OS)
- renamed some gulp tasks to use shorter names
- delete `gulp g3w-admin:client_only` npm script

Raruto committed:
Merge branch 'migration_v3_v4' of https://github.com/Raruto/g3w-client into migration_v3_v4

Raruto committed:
rename index.html templates

Raruto committed:
split i18n into multiple language files

Raruto committed:
install and configure eslint 8

Raruto committed:
apply airbnb styling rules (`eslint src --fix`)

Raruto committed:
remove useless `constructor() { super(); }`

Raruto committed:
remove useless `constructor(options = {}) { super(options); }`

Raruto committed:
remove useless `constructor(config = {}, options = {}) { super(config, options); }`

Raruto committed:
remove useless `constructor(options = {}, extraParams = {}) { super(options, extraParams); }`

Raruto committed:
replace some relative imports  (eg. `../..`)

volterra79 committed:
fiexd issue on missing icon on sidebar components

volterra79 committed:
fixed use of fromExtent Polygon class method on geo.js

volterra79 commented on src/app/gui/component.js:
```diff
@@ -39,10 +38,12 @@ class Component extends G3WObject {
       info = null,
       loading = false,
       disabled = false,
+      icon= null,
```
Pay attention to merge(this, options) on other modules

volterra79 committed:
added print translations

volterra79 committed:
aligned with dev branch set query feature header with link or image icon in case of link or image value attribute

volterra79 committed:
removed coordinates console.log on mapservice.js

volterra79 committed:
fixed issue on wfs provider on querybybbox control request

volterra79 committed:
refactored utils.js splitting default and single function constant export

volterra79 committed:
fixed issue on gulp file using arrow function. 'this' is not referred to bundle but global

volterra79 committed:
refactoried gor.js file splitting default and single constant and functions

volterra79 committed:
fixed issue on missing context on postcompose event in ol6. Temporary workaround. Try to find a better solution

volterra79 committed:
fix issue on parsing xml response in parser.js
const reg = new RegExp(`qgs:${sanitizeLayerName}\\b`, "g"); => const reg = new RegExp(`qgs:${sanitizeLayerName}`, "g");

volterra79 committed:
fix issue get map screenshot and geoscreenshot. Map canvas dom position is changed

volterra79 committed:
fix issue length and area map control

volterra79 committed:
replace headeritem.html and headeritem.js to headeritem vue single component

volterra79 committed:
remove old reference to headeritem.html

volterra79 committed:
remove template attribute on headeritem component

volterra79 committed:
replace lineXY.js and lineXY.html with lineXY.vue vue single component

volterra79 committed:
replace addlayer.js and addlayer.html with addlayer.vue vue single component

volterra79 committed:
change api.js importing lineXY.vue component instead a lineXY.js

volterra79 committed:
refactor form component and its subcomponents footer, body and header and change all references

volterra79 committed:
in case of single language in client select language on header is hide

Raruto committed:
refactor main.js

- remove dead code and comments
- remove useful `bootstrap()` function closure
- simplify error variable parsing
- make use of "ApplicationTemplate" variable instead of reassining default export from `gui/app/index` to `Application`

Raruto committed:
refactor api.js

- remove temporary `g3w` const variable (use default export instead)
- remove unusued `test` property

Raruto committed:
remove useful `production` variable

Raruto committed:
remove useful `G3W_VERSION` constant

Raruto committed:
sync with dev https://github.com/g3w-suite/g3w-client/commit/3e241357e76ef504297b55350dfa686c8914a3d8

Raruto committed:
sync with dev https://github.com/g3w-suite/g3w-client/commit/cc1be1015c3390c681dd54e8900aa4aed9e74692

Raruto committed:
major `version` update in package.json

volterra79 committed:
create component folder inside gui and move all component files inside it.

volterra79 committed:
Merge remote-tracking branch 'origin/migration_v3_v4' into migration_v3_v4

volterra79 committed:
substitute Lodash methods _.forEach with Object.entries().forEach

volterra79 committed:
move all subfolder and files of src/g3w-ol/src inside src/g3w-ol. Remove src/g3w-ol/src

Raruto commented:
Superseded by [g3w-client@v4.0.0](https://github.com/g3w-suite/g3w-client/tree/v4.0.0)
